File: shared/frontmatter.ts
import * as yaml from "js-yaml";

// Shared between the Node server and the browser client — the one thing
// gray-matter (the original server-side choice) can't be, since it pulls
// in Node's Buffer global under the hood and crashed the moment
// PropertiesPanel tried to use it client-side. js-yaml has no such
// dependency, so this single implementation replaces both. (Not shared
// with the Rust/Tauri side — src-tauri/src/frontmatter.rs is a separate
// implementation by necessity of the language boundary; three
// implementations becomes two.)

export interface ParsedFrontmatter {
  data: Record<string, unknown>;
  body: string;
}

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?/;

export function parseFrontmatter(raw: string): ParsedFrontmatter {
  const match = FRONTMATTER_RE.exec(raw);
  if (!match) return { data: {}, body: raw };
  let data: Record<string, unknown> = {};
  try {
    const loaded = yaml.load(match[1]);
    if (loaded && typeof loaded === "object") data = loaded as Record<string, unknown>;
  } catch {
    // Malformed YAML in the frontmatter block — treat as no properties
    // rather than crashing.
  }
  return { data, body: raw.slice(match[0].length) };
}

export function stringifyFrontmatter(data: Record<string, unknown>, body: string): string {
  if (Object.keys(data).length === 0) return body;
  const yamlText = yaml.dump(data).trimEnd();
  return `---\n${yamlText}\n---\n${body}`;
}

File: src-tauri/src/frontmatter.rs
use regex::Regex;
use serde_json::{Map, Value};
use std::sync::OnceLock;

fn frontmatter_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?s)^---\r?\n(.*?)\r?\n---\r?\n?").unwrap())
}

pub struct ParsedFrontmatter {
    pub data: Map<String, Value>,
    pub body: String,
}

/// Same `---`-fenced YAML convention as src/frontmatter.ts (the browser
/// side) and server/vault.ts (gray-matter, in the Node deployment) — three
/// independent implementations of one simple format, kept in sync by hand
/// rather than shared code, since they're in three different languages.
pub fn parse(raw: &str) -> ParsedFrontmatter {
    match frontmatter_re().captures(raw) {
        Some(caps) => {
            let yaml_str = &caps[1];
            let data = serde_yaml::from_str::<Value>(yaml_str)
                .ok()
                .and_then(|v| v.as_object().cloned())
                .unwrap_or_default();
            let body = raw[caps.get(0).unwrap().end()..].to_string();
            ParsedFrontmatter { data, body }
        }
        None => ParsedFrontmatter {
            data: Map::new(),
            body: raw.to_string(),
        },
    }
}

